Changes to ViewFlags for 3.0

pmconne opened the pull request:
Breaking changes for iModel.js 3.0.
See NextVersion.md for details.

pmconne commented on ui/imodel-components/src/ui-imodel-components/navigationaids/DrawingNavigationAid.tsx:
```diff
@@ -763,7 +763,8 @@ export class DrawingNavigationCanvas extends React.Component<DrawingNavigationCa
       vp.logo.style.display = "none";
-    vp.viewFlags.acsTriad = false;
+
+    vp.viewFlags = vp.viewFlags.with("acsTriad", false);
```
@bsteinbk @DanEastBentley can one of you tell me how to satisfy moq for DrawingNavigationAid.test.tsx?
With no changes to the test code, I get 8 failures: "vp.viewFlags.with is not a function".
If I add `vpMock.setup((x) => x.viewFlags).returns(() => new ViewFlags())` I get 8 failures: "trap returned falsish for property".

DanEastBentley replied:
You'll probably need something like:
```ts
const viewFlagsMock = moq.Mock.ofType<ViewFlags>();
viewFlagsMock.setup((vf) => vf.with).returns(() => . . .);  // You'll need to provide implementation
. . .
vp.setup((x) => x.viewFlags).returns(() => viewFlagsMock.object);
```

pmconne replied:
```
      const viewFlags = new ViewFlags();
      const viewFlagsMock = moq.Mock.ofType<ViewFlags>();
      viewFlagsMock.setup((vf) => vf.with).returns((_vf) => (prop, value) => {
        return viewFlags.with(prop, value);
      });

      const vpMock = moq.Mock.ofType<ScreenViewport>();
      vpMock.setup((x) => x.view).returns(() => viewState.object);
      vpMock.setup((x) => x.viewFlags).returns(() => viewFlagsMock.object);
      return vpMock.object;
```
"trap returned falsish for property".

bsteinbk replied:
I am not sure why Paul's solution did not work unless we need to return the same set of flags each time.
```
const testViewFlags = new ViewFlags();
vp.setup((x) => x.viewFlags).returns(() => testViewFlags );
```

pmconne replied:
Tried that one already, @bsteinbk - same "'set' on proxy: trap returned falsish for property 'viewFlags'" error.

DanEastBentley replied:
How large is the ViewFlag.with implementation? Can you include it in the test instead of relying on a real ViewFlags object?

pmconne replied:
It's small, but for this test it could be replaced with a mock that simply returns the same object, or a new one, or whatever - the test shouldn't care.
The problem though appears to be the assignment to `Viewport.viewFlags`, not the call of `with`.
```
  public with(flag: keyof Omit<ViewFlagsProperties, "renderMode">, value: boolean): ViewFlags {
    if (this[flag] === value)
      return this;

    const props: ViewFlagsProperties = { ...this };
    props[flag] = value;
    return new ViewFlags(props);
  }
```

bsteinbk replied:
I am not sure why there are two viewport mocks in these tests,  one named `vp` and one `ScreenViewportMock`.  Did you add viewFlags to both? 

`vp.setup((x) => x.viewFlags).returns(() => testViewFlags );`
and 
`vpMock.setup((x) => x.viewFlags).returns(() => testViewFlags);`

pmconne replied:
I didn't notice that. But the callstack shows the errors happening in this mock, not that other one.
Adding viewFlags to the other one made no difference.

pmconne replied:
I gave up trying to figure out if this is a me problem or a typemoq problem, and just changed customizeViewport to only assign to viewFlags if necessary.
For the test, it is not necessary.

bsteinbk replied:
You will need to add /* istanbul ignore next */ above your new if statement or the test coverage will not be 100% and cause you more grief.

kabentley commented on core/bentley/src/UtilityTypes.ts:
```diff
@@ -1,31 +1,44 @@
+ * @module Utils
+ */
+
+/** The inverse of TypeScript's Readonly<T> type, producing a type that has all the properties of `T` with any `readonly` modifiers removed.
```
this is a nice collection of types. Would it be appropriate to move [this types](https://github.com/imodeljs/imodeljs/blob/e252cfd54429007970838bc995e7ab80e44a8f05/core/frontend/src/IpcApp.ts#L15-L30) here too?

pmconne replied:
Yes, I'll move them.

pmconne replied:
Moved to bentleyjs-core. imodeljs-frontend re-exports them (deprecated), pending decision regarding API deprecations in 3.0.

pmconne commented on test-apps/display-performance-test-app/src/frontend/TestRunner.ts:
```diff
@@ -718,6 +716,15 @@ export class TestRunner {
     if (other)
       testName += `_${other}`;
 
+    // NB: We removed the 3 lighting properties from ViewFlags, replacing them with a single `lighting` property.
```
@DStradley the view flags string will be the last thing in the test name except for the optional test number used to differentiate duplicate test names. That last bit always begins with "---". So your should be able to append "-lit" to any test names that contain "_SmoothShade" and *none* of "+slL", "+cmL", or "+scL".

pmconne commented on test-apps/display-performance-test-app/src/frontend/TestRunner.ts:
```diff
@@ -988,9 +995,11 @@ function getRenderMode(vp: ScreenViewport): string {
   let optString = "";
-  for (const [key, value] of Object.entries(curRenderOpts)) {
+  for (const propName of Object.keys(curRenderOpts)) {
+    const key = propName as NonFunctionPropertyNamesOf<RenderSystem.Options>;
```
@DStradley compiler will now warn us if we delete or change the type of any ViewFlags property. It can't help us if we add a new property.

pmconne commented on test-apps/display-performance-test-app/src/frontend/TestRunner.ts:
```diff
@@ -1028,51 +1037,50 @@ function getRenderOpts(curRenderOpts: RenderSystem.Options): string {
-        if (!value) optString += "-solShd";
+        if (!curRenderOpts[key]) optString += "-solShd";
         break;
-      case "logarithmicZBuffer":
```
@DStradley this was always a bug - wrong property name - compiler caught it now. Inverted this and some others because they are on by default.

pmconne commented on test-apps/display-performance-test-app/src/frontend/TestRunner.ts:
```diff
@@ -1028,51 +1037,50 @@ function getRenderOpts(curRenderOpts: RenderSystem.Options): string {
+        const value = curRenderOpts[key];
+        if (value && value > 1) optString += `+aa${value}`;
         break;
-      default:
```
Many options we have added or will add in future should not show up here. If we want them to, we must add them explicitly.

pmconne commented on test-apps/display-performance-test-app/src/frontend/TestRunner.ts:
```diff
@@ -1028,51 +1037,50 @@ function getRenderOpts(curRenderOpts: RenderSystem.Options): string {
-        if (value) tilePropsStr += `+${key}`;
-    }
-  }
+  if (!curTileProps.enableInstancing)
```
Many of these have no effect on the image. Other properties that weren't being included do affect the image. I changed some to only add to the test name if they differ from their default values.

pmconne commented on test-apps/display-performance-test-app/src/frontend/TestRunner.ts:
```diff
@@ -1146,10 +1154,11 @@ function getOtherProps(vp: ScreenViewport): string {
 function getViewFlagsString(test: TestCase): string {
   let vfString = "";
-  for (const [key, value] of Object.entries(test.viewport.viewFlags)) {
+  const renderMode = test.viewport.viewFlags.renderMode;
```
Several of these (lighting, visible edges, etc) are always true or always false for certain render modes. I omit them if they cannot vary for the given render mode.

pmconne commented:
@DStradley @hnn0003 removal of some view flags and addition of new "lighting" flag cause display-performance-test-app to produce different file names for the same test cases. We'll have to write a script to rename all the previously-generated files. I tried to make doing so as straightforward as possible:
- I kept the loops that iterate `Object.keys` so that the ordering of the abbreviations will remain consistent.
- I updated the code to be type-safe such that if we delete or change the type of a property of ViewFlags, RenderSystem.Options, or TileAdmin.Props in the future, the compiler will produce an error.
- I removed `default` cases that would append the name of any property it didn't recognize. The only such property that shows up in your tests was "thematicDisplay". We now explicitly handle that one. I didn't bother to shorten it.
- I removed the following. Your script should delete any occurrences of them:
	- +shadeSrc, +elide. These don't affect the image.
	- +logZBuf (was never output - property is named logarithmicDepthBuffer but code was checking for logarithmicZBuffer - with my changes, the compiler caught that).
	- +max[N] and +retry[N] where N is an integer. These do not affect the image. Your script should remove them (see regex in `removeOptsFromString`).
	- +scL, +cmL, +slL, +noG, +hln, +genM, +useM. The corresponding ViewFlags properties no longer exist.
- I added -lit to indicate lighting is disabled. I only include this if the render mode is SmoothShade, because it has no effect on the other render modes. If present, it always immediately follows the render mode, separated by an underscore.
	- So, your script can search for +scL, +slL, and +cmL. If present, delete them. Otherwise, search for SmoothShade_ and, if present, insert -lit just after the underscore.

pmconne commented:
@calebmshafer what do you want me to do about deprecations?
- I deleted some obsolete properties from ViewFlags and ViewFlagProps. I can deprecate them on 2.19.x - they all either have replacements, or are useless.
- I removed mutating methods from ViewFlags and made all of its properties read-only. I don't see what I can do on 2.19.x other than mark the removed methods as deprecated saying "this will be removed in 3.0".
- I replaced the ViewFlagOverrides class with a simple interface. I don't see what I can do on 2.19.x other than mark it deprecated saying "this will be replaced in 3.0".
- I moved 3 types from imodeljs-frontend to bentleyjs-core. In this PR, I made the frontend ones re-export the bentleyjs-core ones and marked them deprecated.

I'm not a huge fan of deprecating an API without supplying an immediate replacement.

kabentley commented on core/bentley/src/UtilityTypes.ts:
```diff
@@ -49,3 +49,20 @@ export type NonFunctionPropertyNamesOf<T> = {
+
+/** Any function returning a Promise.
+ * @see [[AsyncMethodsOf]] to extract all async methods from a type.
+ * @see [[PromiseReturnType to extract the type to which the Promise resolves.
```
typo (missing "]]")

pmconne replied:
fixed.

kabentley commented:
> @calebmshafer what do you want me to do about deprecations?
> I'm not a huge fan of deprecating an API without supplying an immediate replacement.

I don't find it alarming to have some newly-deprecated methods in an X.0 release. Why should that be illegal?

pmconne commented:
> I don't find it alarming to have some newly-deprecated methods in an X.0 release. Why should that be illegal?

I don't love that but I'm not calling for it to be outlawed either.
The more important point is, if for example I replace ViewFlagOverrides class with Partial<ViewFlagsProperties> in 3.0, there's nothing for me to mark deprecated in 3.0, and nothing for users to replace their usage with in 2.x if I deprecate it there.

calebmshafer commented:
> > I don't find it alarming to have some newly-deprecated methods in an X.0 release. Why should that be illegal?
> 
> I don't love that but I'm not calling for it to be outlawed either. A big advantage of an X.0 release though is the opportunity to purge legacy baggage. Introducing new legacy baggage with an X.0 release seems undesirable.
> 
> The more important point is, if for example I replace ViewFlagOverrides class with `Partial<ViewFlagsProperties>` in 3.0, there's nothing for me to mark deprecated in 3.0, and nothing for users to replace their usage with in 2.x if I deprecate it there.

@pmconne I tend to agree with @kabentley, I think the general rule for anything that isn't deprecated  in the 2.x version and now has a replacement should be marked as deprecated. However, if the replacement is straight-forward, such as moving a type from "imodeljs-frontend" -> "bentleyjs-core", I think that one should be allowed.  It can also be a part of the "code mod" tool we're working on that would make the transition from 2.x -> 3.x easier.

@skirby1996 is investigating/prototyping that type of tool now. Simple renames and/or moves are definitely in the feature set of the tool we're hoping to build. Anything more complex such as having to change to a new method and call it a different way, is most likely not going to happen unless we find a really good tool that does some of the work for us already.

pmconne commented:
@calebmshafer I am asking for instructions for what, if anything, to do about each of the changes I listed.

kabentley commented:
> @calebmshafer I am asking for instructions for what, if anything, to do about each of the changes I listed.

I suggest you *not* go back and make any 2.19 changes for this. Your clear and complete documentation in NextVersion.md is all anyone will need to adapt their code. I see no reason to worry them about any of this before they upgrade. How would that make any less work for them?

BTW, this PR is a masterpiece, as was your ColorDef changes. It serves as a good example of how immutable types can simplify TS apis. 

Also, is there no way to have a threaded discussion in GitHub other than commenting inside a source file?

pmconne commented:
> I suggest you _not_ go back and make any 2.19 changes for this. Your clear and complete documentation in NextVersion.md is all anyone will need to adapt their code. I see no reason to worry them about any of this before they upgrade. How would that make any less work for them?

I agree, but I gathered @calebmshafer wants anything removed in one release to have been deprecated somehow in a previous release.

> Also, is there no way to have a threaded discussion in GitHub other than commenting inside a source file?

Nope. You also can't resolve comments that aren't attached to source code. But such comments tend to get lost / confusing after changes are made to the associated code.

kabentley commented on core/frontend/src/UtilityTypes.ts:
```diff
@@ -0,0 +1,18 @@
+/*---------------------------------------------------------------------------------------------
```
I'd call this file `DeprecatedTypes.ts` so people will know not to include it in frontend code and that it will go away in 4.0.

pmconne replied:
It needs a packageDocumentation tag. Not all deprecated types will be in the same documentation group. People who include it in frontend code will be told by linter that they are using a deprecated type. I don't want to create a documentation group called "deprecated stuff" in our published docs.

kabentley replied:
i'd still rename it so no one adds anything else to it.

pmconne replied:
done.

mdastous-bentley commented:
@pmconne I looked into it, looks good to me.  Were you expecting breaking changes in the map-layers related stuff?

pmconne commented:
> @pmconne I looked into it, looks good to me. Were you expecting breaking changes in the map-layers related stuff?

Nope.

hnn0003 commented on test-apps/display-performance-test-app/src/frontend/TestRunner.ts:
```diff
@@ -1021,58 +1022,43 @@ function getRenderOpts(curRenderOpts: RenderSystem.Options): string {
               case "EXT_frag_depth":
                 optString += "-fragDepth";
                 break;
-              default:
-                optString += `-${ext}`;
```
@pmconne, you've got a few places in here (this is one) where it's getting rid of all of the defaults, so any new options that are made in core won't automatically get reflected in the test configurations as they have been--you'd have to make an intentional change to dpta itself to add in a new flag for the csv data or png file names. I'm guessing that's intentional, but I thought I should double check.

pmconne replied:
Thanks @hnn0003 - that is intentional. We don't want the filenames to change unexpectedly - if we add a new option that we think should be included in the filename, we should add it explicitly. There are lots of options that have no effect on the image produced.
